src/platform/qt/library/LibraryController.cpp

```diff
@@ -0,0 +1,173 @@
+/* Copyright (c) 2014-2017 waddlesplash
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include "LibraryController.h"
+
+#include "../GBAApp.h"
+#include "LibraryGrid.h"
+#include "LibraryTree.h"
+
+namespace QGBA {
+
+void AbstractGameList::addGames(QList<LibraryEntry*> items) {
+	for (LibraryEntry* o : items) {
+		addGame(o);
+	}
+}
+void AbstractGameList::removeGames(QList<LibraryEntry*> items) {
+	for (LibraryEntry* o : items) {
+		removeGame(o);
+	}
+}
+
+LibraryController::LibraryController(QWidget* parent_widget, const QString& path, ConfigController* config)
+	: QStackedWidget(parent_widget),
+	  m_config(config)
+{
+	if (!path.isNull()) {
+		m_library = mLibraryLoad(path.toUtf8().constData());
+	} else {
+		m_library = mLibraryCreateEmpty();
+	}
+
+	mLibraryAttachGameDB(m_library, GBAApp::app()->gameDB());
+
+	m_libraryTree = new LibraryTree(this);
+	addWidget(m_libraryTree);
+	connect(m_libraryTree, &LibraryTree::StartGame, this, &LibraryController::startGame);
+
+	m_libraryGrid = new LibraryGrid(this);
+	addWidget(m_libraryGrid);
+	connect(m_libraryGrid, &LibraryGrid::StartGame, this, &LibraryController::startGame);
+
+	refresh();
+	selectLastBootedGame(); // TODO: doesn't belong here
+
+	setViewStyle(LibraryStyle::STYLE_TREE);
+}
+
+LibraryController::~LibraryController() {
+	for (LibraryEntry* file : m_games.values()) {
+		delete file;
+	}
+	if (m_library) {
+		mLibraryDestroy(m_library);
+	}
+}
+
+void LibraryController::setViewStyle(LibraryStyle newStyle) {
+	m_currentStyle = newStyle;
+
+	if (newStyle == LibraryStyle::STYLE_LIST || newStyle == LibraryStyle::STYLE_TREE) {
+		m_libraryTree->selectEntry(selectedEntry());
+		setCurrentWidget(m_libraryTree);
+		m_libraryTree->setViewStyle(newStyle);
+	} else {
+		m_libraryGrid->selectEntry(selectedEntry());
+		setCurrentWidget(m_libraryGrid);
+		m_libraryGrid->setViewStyle(newStyle);
+	}
+}
+
+void LibraryController::selectEntry(LibraryEntry* entry) {
+	if (currentWidget() == m_libraryGrid) {
+		m_libraryGrid->selectEntry(entry);
+	} else {
+		m_libraryTree->selectEntry(entry);
+	}
+}
+
+LibraryEntry* LibraryController::selectedEntry() {
+	if (currentWidget() == m_libraryGrid) {
+		return m_libraryGrid->selectedEntry();
+	} else {
+		return m_libraryTree->selectedEntry();
+	}
+}
+
+VFile* LibraryController::selectedVFile() {
+	LibraryEntry* entry = selectedEntry();
+	if (entry) {
+		mLibraryEntry bogus_entry; // TODO: this is probably a bad idea?
```

Comment:
Yes, this is a bad idea. `LibraryEntry` should be a thin wrapper around `mLibraryEntry`. Just store the mLibraryEntry and query data off of it instead of duplicating all of the fields.